Update TransportProtocols.md (#17082)
dotnet/aspnetcore: src/SignalR/docs/specs/TransportProtocols.md

A Poll is established by sending an HTTP GET request to `[endpoint-base]` with the following query string parameters

> ## HTTP Streaming (HTTP/1, HTTP/2, HTTP/3)

Reviewer: So are we committing to some of this for 5.0? @anurse

Reviewer 2: We do need to talk about the scope, but putting it in the spec doesn't *technically* require us to implement it, because of transport negotiation. It just means that when we do implement it, this is how it is done :).

I do think this will fit in the 5.0 milestone, but I also think that can be a somewhat orthogonal discussion.

@davidfowl we should include the transport names (as returned by `/negotiate`) in this doc, for this transport and all the others.

Author: I won't merge it until we decide.

> HTTP Streaming uses a single bi-directional HTTP request or 2 uni-directional HTTP requests as a single connection (one for input, one for output). If the HTTP client in use supports bi-directional streaming, negotiation can be skipped (if the client knows the server supports the streaming transport), otherwise, it requires a connection already be established using the `POST [endpoint-base]/negotiate` request.

Reviewer 2: ```suggestion
HTTP streaming uses [chunked transfer coding](https://tools.ietf.org/html/rfc7230#section-4.1) to stream data between the client and the server on a single HTTP request.

If the client supports *simultaneous* bi-directional streaming. then a single HTTP request is used and negotiation can be skipped (if the client has prior knowledge that the server supports the streaming transport). In order to use this transport, the server must be able to send response body chunks while still receiving request body chunks, and the client must be able to receive response body chunks while still sending request body chunks.

For clients that only support response body streaming (such as browsers), negotiation must be performed in order to allocate a connection ID. A single streaming request is used to send data from server to client (as with Server-Sent Events). The HTTP POST pattern described above is used to send data from client to server.
```

Reviewer 2: Updated my suggestion to remove reference to chunking, since that's only how it's done in HTTP/1.1, not HTTP/2+

> ### 1 bi-directional HTTP Request

Reviewer 2: ```suggestion
### Bi-directional HTTP Request
```

Style nit: Avoid starting headings/sentences/etc. with numerals.

Reviewer 2: If you really want to drive the "one request" home, use

```suggestion
### Single bi-directional HTTP Request
```

> For clients that support bi-directional streaming, the HTTP Streaming transport is activated by making a POST request `[endpoint-base]` with the content type `application/bedrock-streaming`. The **optional** `id` query string value is used to identify the connection to attach to. 

Reviewer 2: `application/bedrock-streaming` Hooooo boy that name ;).

Author: Yea that name sucks

If there is no `id` query string value, a new connection is established. If the parameter is specified but there is no connection with the specified ID value, a `404 Not Found` response is returned. 

Establishing a second streaming connection when there is already a connection associated with the Endpoints connection is not permitted and will fail with a `409 Conflict` status code.

> Errors while establishing the connection are handled by returning a `500 Server Error` status code as the response to the upgrade request. This includes errors initializing EndPoint types. 

Reviewer 3: upgrade request? Was this copied from WebSockets?

Reviewer 2: Busted!

Author: shh

> ### 1 uni-directional HTTP Requests (Server-to-Client only)

Author: I'm torn whether to describe the POST here as well or to treat the long running POST similar to how we treat the short lived POST.

> ### 1 uni-directional HTTP Requests (Server-to-Client only)

Reviewer 2: ```suggestion
### Uni-directional HTTP Requests (Server-to-Client only)
```

Same nit.

> For clients that do not support bi-directional streaming (like browser clients), it requires a connection already be established using the POST `[endpoint-base]/negotiate` request. The HTTP Streaming transport running in this mode is activated by making a GET request `[endpoint-base]` with the content type `application/bedrock-streaming`.

Reviewer 2: How does the server know to allow POSTs because this is the uni-directional version?

Reviewer 3: The POST acts as the upload half of the stream in this case? How does the client know the stream was successfully established?

Reviewer 4: Presumably by getting a 200 response from the "GET request `[endpoint-base]` with the content type `application/bedrock-streaming`". By the time the server responds with the 200 to a get request, the client should be free to POST since at that point the server would not about the client and its unidirectionalness.

Reviewer 2: For SSE (which is very similar) I believe browsers wait until the headers arrive. I expect we'd do the same here (assuming there's some way to observe that in the browser). If not, we may need some kind of "hello" message to be sent down the stream to confirm that it's working.

Reviewer: Having tested this with fetch and streaming there was a need for a "hello" message only on firefox (which was also needed in SSE only on firefox)
